lib/active_merchant/billing/gateways/moneris_us.rb

```diff
@@ -154,14 +157,39 @@ def expdate(creditcard)
         sprintf("%.4i", creditcard.year)[-2..-1] + sprintf("%.2i", creditcard.month)
       end
 
+      def add_address(post, payment_method, options)
+        if !payment_method.is_a?(String) && card_brand(payment_method) == 'check'
+          post[:ach_info][:cust_first_name] = payment_method.first_name if payment_method.first_name
+          post[:ach_info][:cust_last_name] = payment_method.last_name if payment_method.last_name
+          if address = options[:billing_address] || options[:address]
+            post[:ach_info][:cust_address1] = address[:address1] if address[:address1]
+            post[:ach_info][:cust_address2] = address[:address2] if address[:address2]
+            post[:ach_info][:city] = address[:city] if address[:city]
+            post[:ach_info][:state] = address[:state] if address[:state]
+            post[:ach_info][:zip] = address[:zip] if address[:zip]
+          end
+        else
+          post[:address] = options[:billing_address] || options[:address]
+        end
+      end
+
       def add_payment_source(post, source, options)
         if source.is_a?(String)
           post[:data_key]   = source
           post[:cust_id]    = options[:customer]
+        elsif card_brand(source) == 'check'
+          ach_info = {}
+          ach_info[:sec] = 'web'
+          ach_info[:routing_num] = source.routing_number
+          ach_info[:account_num] = source.account_number
+          ach_info[:account_type] = source.account_type
+          ach_info[:check_num] = source.number if source.number
+          post[:ach_info] = ach_info
         else
           post[:pan]        = source.number
           post[:expdate]    = expdate(source)
           post[:cvd_value]  = source.verification_value if source.verification_value?
+          post[:crypt_type] = options[:crypt_type] || @options[:crypt_type] if (options[:crypt_type] || @options[:crypt_type])
```

Comment:
This is 100% up to you, but an alternative to the heavy repetition here would be

    if crypt_type = options[:crypt_type] || @options[:crypt_type]
        post[:crypt_type] = crypt_type
    end

which is _slightly_ longer, but might be easier to read. Again, entirely your call.